### PR TITLE
Fix #6083: rewrite demangleLabels

### LINKED ISSUE #6083: [BUG]: Misplaced label links on demangled inline class members under Clang
### Describe the bug

oddly specific, yes

### Steps to reproduce

- Switch to Clang
- Turn off 'Filter Directives'
- Create a class, a function declared inside the class declaration (for example a constructor), and something that calls that function (for example a global variable of that type)
- Scroll down to that function's .section directive in the output (in the repro link, it's line 27)

### Expected behavior

The 'follow label' link is placed on the demangled symbol

### Observed output

The link is (demangled symbol name length, minus mangled symbol name length) characters too far left

### Reproduction link

https://godbolt.org/z/fz31fs498

### Screenshots

image

### Operating System

_No response_

### Browser version

_No response_

### PR DESCRIPTION
**Pull request by OfekShilon**
This code had some problems:
https://github.com/compiler-explorer/compiler-explorer/blob/07d5c3312fce98d2742fa7ed738d4084f3a6975f/lib/demangler/base.ts#L200-L203
There

`PrefixTree.replaceAll` replaces mangled to demangled names 'blindly', with no notion of labels. Later `demangleLabels` runs separate logic to update the column-range of each label in hope of mimicking the `replaceAll` action.  This fails in some cases: 
(1) the `demangleLables` code handles only a single occurrence of each label in a line, 
(2) It also diverges from `replaceAll` in case that a string which is a label-name appears in the line *not* as part of a label (this does happen!).

This fix makes `replaceAll` return detailed account of its actions, and then use these to properly adjust the label ranges.

This fix might not address all the root issues, but it does make some `label` decisions consistent where they weren't.

### CODE REVIEW

File: lib/demangler/prefix-tree.ts [hunk 1]
```diff
@@ -86,25 +91,51 @@ export class PrefixTree {
     }
 
     // Replace all matches (longest match first) in a line.
-    replaceAll(line: string) {
-        let result = '';
-        let index = 0;
+    replaceAll(line: string): {
```
Comment (partouf) [inline lib/demangler/prefix-tree.ts:94]: can you make this return type an explicit type

**Comment by partouf** [comment]
Looks solid, just needs fixing of tests (maybe also add a test for a situation in which it would previously would've gone wrong)

**Comment by OfekShilon** [comment]
> Looks solid, just needs fixing of tests (maybe also add a test for a situation in which it would previously would've gone wrong)

Sorry, I wasn't able to add a test for this within reasonable effort. Maybe in the future. All else is done.

**Comment by OfekShilon** [comment]
BTW this bug was *exceptionally* hairy and exposed some other worm-cans around label handling. Note that for the offending asm line in the bug,
```
  .section   .text._ZN1aC2Ev,"axG",@progbits,_ZN1aC2Ev,comdat
```
`getUsedLabelsInLine` returned:
image
Note in particular the difference between `_ZN1aC2Ev` and `._ZN1aC2Ev`. 

They're detected separately due to the eagerness of the used regex `this.identifierFindRe = /[$.@A-Z_a-z]\w*/g;`. 
Later the one starting with `.` is ommitted by `removeLabelsWithoutDefinition`.
Still later, PrefixTree.replaceAll (which has no notion of labels) replaces all occurrences of `_ZN1aC2Ev` with the demangled version, including the ones dropped earlier such as `._ZN1aC2Ev`. 

I *think* in this particular example we may have just gotten away with it, because the first arg to `.section` is `.text._ZN1aC2Ev` which really isn't the ctor label. So the right thing to do is probably to demangle it but not link it to the ctor (which is exactly what the code after this PR does). 

But all in all this is the kind of code that "succeeds by chance". I'm not quite happy with it.

**Comment by OfekShilon** [comment]
Perhaps the code can somehow be made to distinguish `label` from `identifier` - currently `identifierFindRe` is used for both, but mangled names cannot start with `.` - only `_` - so more refined regexes could be put into place.   But that is still a half-baked thought.

**Comment by partouf** [comment]
this is now live

**Pull request merged**